File: programmers/타겟 넘버.py
result = 0
list_num = []
visited = []

def dfs(ix,count, cur_val,target) :
    
    if count == len(list_num) :
        global result
        if cur_val == target:
            result += 1
        return

    for i in range(ix,len(list_num)) :

        if visited[i] == False :
            visited[i] == True
            dfs(i+1,count+1, cur_val-list_num[i],target)
            dfs(i+1,count+1, cur_val+list_num[i],target)
            visited[i] = False

def solution(numbers, target):

    for num in numbers :
        list_num.append(num)
        visited.append(False)
    
    dfs(0,0,0,target)
    
    return result

File: programmers/코테대비(2022.01.09)/타겟 넘버.java
class Solution {
    
    static int[] arr;
    static int ans = 0;
    static int dap = 0;
    
    static void dfs(int ix, int count, int curVal){
        if(count == arr.length){
            if(curVal == dap) {
                ans += 1;
            }
            return;
        }
        for(int i=ix;i<arr.length;i++){
            dfs(i+1,count+1,curVal+arr[i]);
            dfs(i+1,count+1,curVal-arr[i]);
        }
    }
    public int solution(int[] numbers, int target) {
        int answer = 0;
        dap = target;
        arr = new int[numbers.length];
        for(int i=0;i<numbers.length;i++){
            arr[i] = numbers[i];
        }
        
        dfs(0,0,0);
        return ans;
    }
}
